--- a/target/linux/mediatek/image/filogic.mk
+++ b/target/linux/mediatek/image/filogic.mk
@@ -12,7 +12,7 @@
 
 define Build/fit-with-netgear-top-level-rootfs-node
 	$(call Build/fit-its,$(1))
-	$(TOPDIR)/scripts/gen_netgear_rootfs_node.sh $(KERNEL_BUILD_DIR)/root.squashfs$(if $(TARGET_PER_DEVICE_ROOTFS),+pkg=$(ROOTFS_ID/$(DEVICE_NAME))) > $@.rootfs
+	$(TOPDIR)/scripts/gen_netgear_rootfs_node.sh $(KERNEL_BUILD_DIR)/root.squashfs > $@.rootfs
 	awk '/configurations/ { system("cat $@.rootfs") } 1' $@.its > $@.its.tmp
 	@mv -f $@.its.tmp $@.its
 	@rm -f $@.rootfs
@@ -81,6 +81,15 @@
 	rm $@.tmp
 endef
 
+define Build/mt798x-gpt-nvme
+	cp $@ $@.tmp 2>/dev/null || true
+	ptgen -g -o $@.tmp -a 1 -l 1024 \
+			-t 0x83 -N boot		-r	-p 63M@1M \
+			-t 0x2e -N production		-p $(CONFIG_TARGET_ROOTFS_PARTSIZE)M@64M
+	cat $@.tmp >> $@
+	rm $@.tmp
+endef
+
 # Variation of the normal partition table to account
 # for factory and mfgdata partition
 #
@@ -391,8 +400,6 @@
 define Device/asus_rt-ax52
   DEVICE_VENDOR := ASUS
   DEVICE_MODEL := RT-AX52
-  DEVICE_ALT0_VENDOR := ASUS
-  DEVICE_ALT0_MODEL := RT-AX52 PRO
   DEVICE_DTS := mt7981b-asus-rt-ax52
   DEVICE_DTS_DIR := ../dts
   DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware
@@ -663,30 +670,70 @@
   DEVICE_VENDOR := Bananapi
   DEVICE_DTS_DIR := $(DTS_DIR)/
   DEVICE_DTS_LOADADDR := 0x45f00000
-  DEVICE_DTS_OVERLAY:= mt7988a-bananapi-bpi-r4-emmc \
-		       mt7988a-bananapi-bpi-r4-rtc \
-		       mt7988a-bananapi-bpi-r4-sd \
-		       mt7988a-bananapi-bpi-r4-spim-nand \
-		       mt7988a-bananapi-bpi-r4-spim-nand-nmbm \
-		       mt7988a-bananapi-bpi-r4-wifi-be14
+  DEVICE_DTS_OVERLAY:= \
+mt7988a-bananapi-bpi-r4-emmc \
+mt7988a-bananapi-bpi-r4-rtc \
+mt7988a-bananapi-bpi-r4-sd \
+mt7988a-bananapi-bpi-r4-spim-nand \
+mt7988a-bananapi-bpi-r4-spim-nand-nmbm \
+mt7988a-bananapi-bpi-r4-nvme
   DEVICE_DTC_FLAGS := --pad 4096
   DEVICE_PACKAGES := kmod-hwmon-pwmfan kmod-i2c-mux-pca954x kmod-eeprom-at24 \
-		     kmod-rtc-pcf8563 kmod-sfp kmod-phy-aquantia kmod-usb3 e2fsprogs f2fsck mkf2fs mt7988-wo-firmware
+    kmod-rtc-pcf8563 kmod-sfp kmod-phy-aquantia kmod-usb3 e2fsprogs f2fsck mkf2fs mt7988-wo-firmware
   DEVICE_COMPAT_VERSION := 1.1
   DEVICE_COMPAT_MESSAGE := The non-switch ports were renamed to match the board/case labels
   KERNEL_LOADADDR := 0x46000000
+
+  ARTIFACTS := \
+      emmc-gpt.bin emmc-preloader.bin emmc-bl31-uboot.fip \
+      emmc-img.bin \
+      nvme-img.bin \
+      sdcard.img.gz \
+      snand-preloader.bin snand-bl31-uboot.fip \
+      snand-img.bin
+  ARTIFACT/emmc-gpt.bin := mt798x-gpt emmc
+  ARTIFACT/emmc-preloader.bin := mt7988-bl2 emmc-ddr4
+  ARTIFACT/emmc-bl31-uboot.fip := mt7988-bl31-uboot $$(DEVICE_NAME)-emmc
+  ARTIFACT/emmc-img.bin := mt798x-gpt emmc | \
+  pad-to 17k | mt7988-bl2 emmc-ddr4 | \
+  pad-to 6656k | mt7988-bl31-uboot $$(DEVICE_NAME)-emmc | \
+  pad-to 64M | append-image squashfs-sysupgrade.itb
+  ARTIFACT/nvme-img.bin := mt798x-gpt-nvme | \
+  pad-to 512M | append-image squashfs-sysupgrade.itb
+  ARTIFACT/snand-preloader.bin := mt7988-bl2 spim-nand-ddr4
+  ARTIFACT/snand-bl31-uboot.fip := mt7988-bl31-uboot $$(DEVICE_NAME)-snand
+  ARTIFACT/snand-img.bin := mt7988-bl2 spim-nand-ddr4 | \
+  pad-to 2048k | \
+  ubinize-image fit squashfs-sysupgrade.itb
+  ARTIFACT/sdcard.img.gz := mt798x-gpt sdmmc |\
+  pad-to 17k | mt7988-bl2 sdmmc-ddr4 |\
+  pad-to 6656k | mt7988-bl31-uboot $$(DEVICE_NAME)-sdmmc |\
+  pad-to 44M | mt7988-bl2 spim-nand-ddr4 |\
+  pad-to 45M | mt7988-bl31-uboot $$(DEVICE_NAME)-snand |\
+  pad-to 51M | mt7988-bl2 emmc-ddr4 |\
+  pad-to 52M | mt7988-bl31-uboot $$(DEVICE_NAME)-emmc |\
+  pad-to 56M | mt798x-gpt emmc |\
+$(if $(CONFIG_TARGET_ROOTFS_SQUASHFS),\
+  pad-to 64M | append-image squashfs-sysupgrade.itb | check-size |\
+) \
+  gzip
+
   KERNEL := kernel-bin | gzip
   KERNEL_INITRAMFS := kernel-bin | lzma | \
-	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
+fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
   IMAGES := sysupgrade.itb
-  KERNEL_INITRAMFS_SUFFIX := -recovery.itb
-  KERNEL_IN_UBI := 1
+
   IMAGES := sysupgrade.itb
   IMAGE/sysupgrade.itb := append-kernel | fit gzip $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-with-rootfs | pad-rootfs | append-metadata
 ifeq ($(DUMP),)
   IMAGE_SIZE := $$(shell expr 64 + $$(CONFIG_TARGET_ROOTFS_PARTSIZE))m
 endif
 
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  UBOOTENV_IN_UBI := 1
+  UBINIZE_PARTS := fip=:$(STAGING_DIR_IMAGE)/mt7988_bananapi_bpi-r4-snand-u-boot.fip
+
 endef
 
 define Device/bananapi_bpi-r4
@@ -1550,38 +1597,18 @@
 endef
 TARGET_DEVICES += huasifei_wh3000
 
-define Device/huasifei_wh3000-pro-emmc
+define Device/huasifei_wh3000-pro
   DEVICE_VENDOR := Huasifei
   DEVICE_MODEL := WH3000 Pro
-  DEVICE_VARIANT := eMMC
-  DEVICE_DTS := mt7981b-huasifei-wh3000-pro-emmc
+  DEVICE_DTS := mt7981b-huasifei-wh3000-pro
   DEVICE_DTS_DIR := ../dts
-  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware \
-	kmod-hwmon-pwmfan kmod-usb3 f2fsck mkf2fs
-  SUPPORTED_DEVICES += huasifei,wh3000-pro
+  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware kmod-hwmon-pwmfan kmod-usb3 f2fsck mkf2fs
   KERNEL := kernel-bin | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
   KERNEL_INITRAMFS := kernel-bin | lzma | \
 	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
-TARGET_DEVICES += huasifei_wh3000-pro-emmc
-
-define Device/huasifei_wh3000-pro-nand
-  DEVICE_VENDOR := Huasifei
-  DEVICE_MODEL := WH3000 Pro
-  DEVICE_VARIANT := NAND
-  DEVICE_DTS := mt7981b-huasifei-wh3000-pro-nand
-  DEVICE_DTS_DIR := ../dts
-  UBINIZE_OPTS := -E 5
-  BLOCKSIZE := 128k
-  PAGESIZE := 2048
-  IMAGE_SIZE := 229376k
-  KERNEL_IN_UBI := 1
-  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
-  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware \
-	kmod-usb3 kmod-hwmon-pwmfan
-endef
-TARGET_DEVICES += huasifei_wh3000-pro-nand
+TARGET_DEVICES += huasifei_wh3000-pro
 
 define Device/imou_hx21
   DEVICE_VENDOR := Imou
@@ -1769,32 +1796,6 @@
 endef
 TARGET_DEVICES += kebidumei_ax3000-u22
 
-define Device/keenetic_kap-630-common
-  DEVICE_DTS_DIR := ../dts
-  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7981-firmware mt7981-wo-firmware \
-		kmod-phy-airoha-en8811h
-  UBINIZE_OPTS := -E 5
-  BLOCKSIZE := 128k
-  PAGESIZE := 2048
-  KERNEL_SIZE := 6144k
-  IMAGE_SIZE := 108544k
-  KERNEL := kernel-bin | lzma | fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb | \
-	append-squashfs4-fakeroot
-  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
-  IMAGES += factory.bin
-  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | \
-	append-ubi | check-size | zyimage -d $$(ZYIMAGE_ID) -v "$$(DEVICE_MODEL)"
-endef
-
-define Device/keenetic_kap-630
-  DEVICE_VENDOR := Keenetic
-  DEVICE_MODEL := KAP-630
-  DEVICE_DTS := mt7981b-keenetic-kap-630
-  ZYIMAGE_ID := 0x810630
-  $(call Device/keenetic_kap-630-common)
-endef
-TARGET_DEVICES += keenetic_kap-630
-
 define Device/keenetic_kn-1812-common
   DEVICE_DTS_DIR := ../dts
   DEVICE_PACKAGES := kmod-mt7992-firmware kmod-usb3 \
@@ -2197,15 +2198,6 @@
 endef
 TARGET_DEVICES += netcore_n60-pro
 
-define Device/netcraze_nap-630
-  DEVICE_VENDOR := Netcraze
-  DEVICE_MODEL := NAP-630
-  DEVICE_DTS := mt7981b-netcraze-nap-630
-  ZYIMAGE_ID := 0xC10630
-  $(call Device/keenetic_kap-630-common)
-endef
-TARGET_DEVICES += netcraze_nap-630
-
 define Device/netcraze_nc-1812
   DEVICE_VENDOR := Netcraze
   DEVICE_MODEL := NC-1812
@@ -3275,31 +3267,3 @@
   IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
 TARGET_DEVICES += zyxel_nwa50ax-pro
-
-define Device/zyxel_wx5600-t0-ubootmod
-  DEVICE_VENDOR := Zyxel
-  DEVICE_MODEL := WX5600-T0
-  DEVICE_VARIANT := (OpenWrt U-Boot layout)
-  DEVICE_DTS := mt7986b-zyxel-wx5600-t0-ubootmod
-  DEVICE_DTS_DIR := ../dts
-  DEVICE_PACKAGES := kmod-mt7915e kmod-mt7986-firmware mt7986-wo-firmware
-  KERNEL_INITRAMFS_SUFFIX := -recovery.itb
-  IMAGES := sysupgrade.itb
-  BLOCKSIZE := 256k
-  PAGESIZE := 4096
-  KERNEL_IN_UBI := 1
-  UBOOTENV_IN_UBI := 1
-  KERNEL := kernel-bin | lzma
-  KERNEL_INITRAMFS := kernel-bin | lzma | \
-        fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd
-  IMAGE/sysupgrade.itb := append-kernel | \
-        fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-static-with-rootfs | append-metadata
-  ARTIFACTS := preloader.bin bl31-uboot.fip
-  ARTIFACT/preloader.bin := mt7986-bl2 spim-nand-4k-ddr3
-  ARTIFACT/bl31-uboot.fip := mt7986-bl31-uboot zyxel_wx5600-t0
-ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
-  ARTIFACTS += initramfs-factory.ubi
-  ARTIFACT/initramfs-factory.ubi := append-image-stage initramfs-recovery.itb | ubinize-kernel
-endif
-endef
-TARGET_DEVICES += zyxel_wx5600-t0-ubootmod
